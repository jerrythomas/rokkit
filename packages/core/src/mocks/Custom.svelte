<script>
	export let value
	export let fields
</script>

<span>{value[fields.text]}</span>
