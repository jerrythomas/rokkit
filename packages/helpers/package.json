{
  "name": "@rokkit/helpers",
  "version": "1.0.0-next.116",
  "description": "Custom matchers for vitest, mocks and simulators for testing.",
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "vitest",
    "matchers",
    "validators",
    "svelte"
  ],
  "author": "Jerry Thomas<me@jerrythomas.name>",
  "license": "MIT",
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    "./mocks": "./src/mocks/index.js",
    "./matchers": "./src/matchers/index.js",
    "./simulators": "./src/simulators/index.js",
    "./components": "./src/components/index.js",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./src/index.js",
			"svelte": "./src/index.js"
    }
  },
  "scripts": {
    "prepublishOnly": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "build": "bun clean && bun prepublishOnly"
  },
  "dependencies": {
    "@vitest/expect": "^2.1.8",
    "ramda": "^0.30.1"
  }
}
