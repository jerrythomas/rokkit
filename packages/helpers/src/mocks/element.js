import { vi } from 'vitest'

/**
 * Creates an array of elements with the specified size
 *
 * @param {number} count
 * @param {number} size
 * @param {string} [prop='offsetHeight']
 * @returns {Array<Object<string, number>>}
 */
export function elementsWithSize(count, size, prop = 'offsetHeight') {
	return Array.from({ length: count }, () => ({
		[prop]: size
	}))
}

/**
 * Creates an array of elements with mixed sizes
 *
 * @param {Array<{count: number, size: number}>} data
 * @param {string} prop
 * @returns {Array<Object<string, number>>}
 */
export function mixedSizeElements(data, prop) {
	return data.reduce(
		(elements, { count, size }) => [...elements, ...elementsWithSize(count, size, prop)],
		[]
	)
}

/**
 * Creates a mock node with functions to add and remove event handlers
 *
 * @param {Array<string} events
 * @returns {{node: HTMLElement, listeners: Object<string, integer>}}
 */
export function getMockNode(events) {
	const listeners = events.reduce((acc, event) => ({ ...acc, [event]: 0 }), {})

	const node = {
		dispatchEvent: vi.fn(),
		scrollTo: vi.fn(),
		querySelectorAll: vi.fn().mockImplementation((selector) => {
			return [document.createElement(selector)]
		}),
		querySelector: vi.fn().mockImplementation((selector) => {
			return document.createElement(selector)
		}),
		addEventListener: vi.fn().mockImplementation((name) => ++listeners[name]),
		removeEventListener: vi.fn().mockImplementation((name) => --listeners[name])
	}
	return { node, listeners }
}

/**
 * @typedef {Object} NestedItem
 * @property {string} name
 * @property {string} [dataPath]
 * @property {string} [id]
 * @property {Array<NestedItem>} [children]
 */

/**
 * Creates a nested HTML element structure using the provided data
 *
 * @param {NestedItem} item
 * @returns {HTMLElement}
 */
export function createNestedElement(item) {
	const { name, dataPath, id, children } = item
	const element = document.createElement(name)

	if (dataPath) element.dataset.path = dataPath
	if (id) element.id = id

	element.scrollIntoView = vi.fn()

	if (Array.isArray(children)) {
		children.forEach((child) => element.appendChild(createNestedElement(child)))
	}

	return element
}

/**
 * Mocks the requestSubmit function for HTMLFormElement which is not implemented by JSDOM
 */
export function mockFormRequestSubmit() {
	// eslint-disable-next-line no-undef
	HTMLFormElement.prototype.requestSubmit = vi.fn(function (submitter) {
		const event = new Event('submit', {
			bubbles: true,
			cancelable: true
		})
		if (submitter) {
			event.submitter = submitter
		}
		this.dispatchEvent(event)
	})

	return true
}
