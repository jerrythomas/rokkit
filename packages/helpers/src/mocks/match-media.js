import { vi } from 'vitest'

const watchMediaQueries = []
const listeners = []

/**
 * @typedef {Object} MediaQuery
 * @property {integer} [min-width]
 * @property {integer} [max-width]
 * @property {integer} [min-height]
 * @property {integer} [max-height]
 * @property {integer} [width]
 * @property {integer} [height]
 * @property {integer} [orientation]
 * @property {integer} [aspect-ratio]
 * @property {integer} [min-aspect-ratio]
 * @property {integer} [max-aspect-ratio]
 * @property {integer} [resolution]
 * @property {integer} [min-resolution]
 * @property {integer} [max-resolution]
 * @property {integer} [scan]
 * @property {integer} [grid]
 * @property {integer} [update]
 * @property {integer} [overflow-block]
 */

/**
 * Parses a media query string into an object
 *
 * @param {string} mediaQuery
 * @returns {MediaQuery}
 */
function parseMediaQuery(mediaQuery) {
	const regex = /\(([^:]+):\s*([^)]+)\)/g
	const result = {}

	let match = null
	while ((match = regex.exec(mediaQuery)) !== null) {
		const [, property, value] = match
		result[property.trim()] = parseInt(value.trim(), 10)
	}

	return result
}

/**
 * Simulates the evaluation of a media query
 *
 * @param {MediaQuery} mediaQuery
 * @param {integer} width
 * @returns
 */
function evaluateMediaQuery(mediaQuery, width) {
	const { 'min-width': minWidth = 0, 'max-width': maxWidth = width } = mediaQuery

	return width >= minWidth && width <= maxWidth
}

/**
 * Mocks the window.matchMedia function
 * @param {string} query
 * @returns {Object}
 */
export const matchMediaMock = vi.fn().mockImplementation((query) => {
	const mediaQuery = parseMediaQuery(query)
	const handler = (width) => evaluateMediaQuery(mediaQuery, width)

	const queryObject = {
		media: query,
		matches: handler(window.innerWidth),
		addListener: vi.fn().mockImplementation((listener) => listeners.push(listener)),
		removeListener: vi
			.fn()
			.mockImplementation((listener) => listeners.splice(listeners.indexOf(listener), 1)),
		handler
	}
	watchMediaQueries.push(queryObject)
	return watchMediaQueries[watchMediaQueries.length - 1]
})

/**
 * Updates the media query matches
 * @returns {void}
 */
export function updateMedia() {
	watchMediaQueries.forEach((query) => {
		query.matches = query.handler(window.innerWidth)
	})
	listeners.forEach((listener) => listener())
}
