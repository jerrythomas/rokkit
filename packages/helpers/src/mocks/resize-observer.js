export class ResizeObserver {
	constructor(callback) {
		this.callback = callback
		this.elements = new Map()
	}

	observe(element) {
		this.elements.set(element, { contentRect: element.getBoundingClientRect() })
		// Immediately invoke the callback with the initial size
		this.callback([{ target: element, contentRect: element.getBoundingClientRect() }])
	}

	unobserve(element) {
		this.elements.delete(element)
	}

	disconnect() {
		this.elements.clear()
	}

	// Simulate a resize event
	simulateResize(element, contentRect) {
		this.callback([{ target: element, contentRect }])
	}
}
