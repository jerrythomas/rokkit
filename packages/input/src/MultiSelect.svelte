<script>
	import Select from './Select.svelte'
	import { Pill } from '@rokkit/core'

	let className = ''
	export { className as class }
	export let value = []
	export let items = []
	export let fields = {}
	export let using = {}

	let available = items

	function handleRemove(event) {
		value = [...value.filter((item) => item == event.detail)]
		available = [...available, event.detail]
	}
	function handleSelect(event) {
		value = [...value, event.detail]
		available = [...available.filter((item) => item == event.detail)]
	}
</script>

<Select
	options={available}
	{fields}
	{using}
	searchable
	on:select={handleSelect}
	class={className}
>
	<items class="flex flex-wrap">
		{#each value as item}
			<Pill {item} {fields} remove on:remove={handleRemove} />
		{/each}
	</items>
</Select>
