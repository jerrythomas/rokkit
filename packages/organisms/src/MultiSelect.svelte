<script>
	import Select from './Select.svelte'
	import { ItemWrapper, Item } from '@rokkit/molecules'
	import { defaultFields } from '@rokkit/core'

	let className = ''
	export { className as class }
	export let value = []
	export let options = []
	/** @type {import('@rokkit/core').FieldMapping} */
	export let fields = {}
	export let using = {}
	export let placeholder = ''

	$: available = options.filter((item) => !value.includes(item))

	function handleRemove(event) {
		value = [...value.filter((item) => item != event.detail)]
	}
	function handleSelect(event) {
		if (!value.includes(event.detail)) value = [...value, event.detail]
	}
	$: using = { default: Item, ...using }
	$: fields = { ...defaultFields, ...fields }
</script>

<Select
	options={available}
	{fields}
	{using}
	on:select={handleSelect}
	class={className}
	{placeholder}
>
	{#if value.length > 0}
		<items class="flex flex-wrap">
			{#each value as item}
				<ItemWrapper
					value={item}
					{fields}
					{using}
					removable
					on:remove={handleRemove}
					class="pill"
				/>
			{/each}
		</items>
	{:else}
		<item class="w-full flex">
			<svelte:component this={using.default} value={placeholder} />
		</item>
	{/if}
</Select>
