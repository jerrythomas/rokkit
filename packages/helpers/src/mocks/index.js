import './animate'
import { ResizeObserver } from './resize-observer'

// skipcq: JS-E1004 - Needed for exposing all functions
export * from './match-media'
// skipcq: JS-E1004 - Needed for exposing all functions
export * from './element'

global.ResizeObserver = ResizeObserver
