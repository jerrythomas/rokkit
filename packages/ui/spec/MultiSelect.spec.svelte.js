import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, fireEvent, render } from '@testing-library/svelte'
import { flushSync, tick } from 'svelte'
import MultiSelect from '../src/MultiSelect.svelte'
// import { MockItem } from '@rokkit/helpers/components'
import '@rokkit/helpers/mocks'

describe('MultiSelect.svelte', () => {
	beforeEach(() => cleanup())

	it('should render using default field mapping', async () => {
		const props = $state({
			name: 'test',
			options: ['One', 'Two', 'Three'],
			value: ['One']
		})
		const { container } = render(MultiSelect, { props })
		expect(container).toBeTruthy()
		const wrapper = container.querySelector('input-select')
		expect(wrapper).toMatchSnapshot()
		const select = wrapper.querySelector('selected-item')
		await fireEvent.click(select)
		await tick()
		expect(wrapper).toMatchSnapshot()

		props.value = ['One', 'Two']
		flushSync()
		await tick()
		expect(wrapper).toMatchSnapshot()
	})

	// it('should render using field mappings', async () => {
	// 	const options = [{ name: 'One' }, { name: 'Two' }, { name: 'Three' }]
	// 	const props = $state({
	// 		options,
	// 		name: 'test',
	// 		value: [options[1]],
	// 		fields: { text: 'name' }
	// 	})
	// 	const { container } = render(MultiSelect, { props })
	// 	expect(container).toBeTruthy()
	// 	const wrapper = container.querySelector('input-select')
	// 	expect(wrapper).toMatchSnapshot()
	// 	const select = wrapper.querySelector('selected-item')
	// 	await fireEvent.click(select)
	// 	await tick()
	// 	expect(wrapper).toMatchSnapshot()

	// 	props.value = [options[3], options[1]]
	// 	flushSync()
	// 	await tick()
	// 	expect(wrapper).toMatchSnapshot()
	// })

	// it('should render items using custom component', async () => {
	// 	const options = [{ name: 'Alpha' }, { name: 'Beta' }, { name: 'Charlie' }]
	// 	const { container } = render(MultiSelect, {
	// 		props: {
	// 			options,
	// 			name: 'test',
	// 			value: [options[0]],
	// 			fields: { text: 'name' },
	// 			using: { default: MockItem }
	// 		}
	// 	})
	// 	expect(container).toBeTruthy()
	// 	const wrapper = container.querySelector('input-select')
	// 	const select = wrapper.querySelector('selected-item')
	// 	await fireEvent.click(select)
	// 	await tick()
	// 	// wrapper = container.querySelector('input-select')
	// 	expect(wrapper).toMatchSnapshot()
	// })

	// it('should add a selected item', async () => {
	// 	const { container, component } = render(MultiSelect, {
	// 		props: {
	// 			name: 'test',
	// 			options: ['One', 'Two', 'Three'],
	// 			value: ['One']
	// 		}
	// 	})
	// 	const wrapper = container.querySelector('input-select')
	// 	const select = wrapper.querySelector('selected-item')

	// 	const selectedItems = select.querySelectorAll('wrap-item')
	// 	expect(selectedItems.length).toBe(1)
	// 	await fireEvent.click(select)
	// 	await tick()

	// 	const items = wrapper.querySelectorAll('scroll item')
	// 	expect(items.length).toBe(2)
	// 	await fireEvent.click(items[1])
	// 	await tick()

	// 	expect(getPropertyValue(component, 'value')).toEqual(['One', 'Three'])
	// 	await fireEvent.click(select)
	// 	await tick()
	// 	expect(wrapper).toMatchSnapshot()
	// })
	// it('should remove a selected item', async () => {
	// 	const { container, component } = render(MultiSelect, {
	// 		props: {
	// 			name: 'test',
	// 			options: ['One', 'Two', 'Three'],
	// 			value: ['One', 'Two']
	// 		}
	// 	})
	// 	const wrapper = container.querySelector('input-select')
	// 	const select = wrapper.querySelector('selected-item')

	// 	const items = select.querySelectorAll('wrap-item')
	// 	expect(items.length).toBe(2)
	// 	await fireEvent.click(items[0].querySelector('icon'))
	// 	await tick()
	// 	expect(getPropertyValue(component, 'value')).toEqual(['Two'])
	// 	await fireEvent.click(select)
	// 	await tick()
	// 	expect(wrapper).toMatchSnapshot()
	// })

	// it('should render with custom class', async () => {
	// 	const { container, component } = render(MultiSelect, {
	// 		props: {
	// 			name: 'test',
	// 			options: ['One', 'Two', 'Three'],
	// 			value: ['One'],
	// 			class: 'custom-class'
	// 		}
	// 	})
	// 	expect(container).toBeTruthy()
	// 	let classes = Array.from(container.querySelector('input-select').classList)
	// 	expect(classes).toContain('custom-class')
	// 	component.$set({ class: 'custom-class-2' })
	// 	await tick()
	// 	classes = Array.from(container.querySelector('input-select').classList)
	// 	expect(classes).not.toContain('custom-class')
	// 	expect(classes).toContain('custom-class-2')
	// })
})
