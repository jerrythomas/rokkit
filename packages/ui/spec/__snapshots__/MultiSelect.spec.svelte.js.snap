// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiSelect.svelte > should render using default field mapping 1`] = `
<input-select
  aria-label="test"
  class="relative flex flex-col "
  role="listbox"
  tabindex="0"
>
  <selected-item
    aria-selected="false"
    class="flex w-full items-center"
    role="option"
    tabindex="-1"
  >
    <!---->
    <items
      class="flex flex-wrap"
    >
      <!---->
      <!---->
      
      <!---->
       
      <p>
        One
      </p>
      <!---->
      
    </items>
    
    <!---->
     
    <rk-icon
      aria-label="closed"
      class=""
      data-tag="icon"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-closed"
      />
    </rk-icon>
    <!---->
  </selected-item>
   
  <!---->
</input-select>
`;

exports[`MultiSelect.svelte > should render using default field mapping 2`] = `
<input-select
  aria-label="test"
  class="relative flex flex-col open"
  role="listbox"
  tabindex="0"
>
  <selected-item
    aria-selected="false"
    class="flex w-full items-center"
    role="option"
    tabindex="-1"
  >
    <!---->
    <items
      class="flex flex-wrap"
    >
      <!---->
      <!---->
      
      <!---->
       
      <p>
        One
      </p>
      <!---->
      
    </items>
    
    <!---->
     
    <rk-icon
      aria-label="opened"
      class=""
      data-tag="icon"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-opened"
      />
    </rk-icon>
    <!---->
  </selected-item>
   
  <scroll
    class="absolute z-10 flex w-full flex-col overflow-scroll menu"
    style="top: 0px;"
  >
    <rk-list
      aria-label="list"
      class=""
      role="listbox"
      tabindex="-1"
    >
      <!---->
       
      <rk-body>
        <!---->
        <rk-list-item
          aria-current="false"
          aria-selected="false"
          data-path="0"
          role="option"
        >
          <!---->
          <!---->
          
          <!---->
           
          <p>
            Two
          </p>
          <!---->
          <!---->
        </rk-list-item>
        <rk-list-item
          aria-current="false"
          aria-selected="false"
          data-path="1"
          role="option"
        >
          <!---->
          <!---->
          
          <!---->
           
          <p>
            Three
          </p>
          <!---->
          <!---->
        </rk-list-item>
        
        <!---->
      </rk-body>
       
      <!---->
    </rk-list>
    <!---->
  </scroll>
  <!---->
</input-select>
`;

exports[`MultiSelect.svelte > should render using default field mapping 3`] = `
<input-select
  aria-label="test"
  class="relative flex flex-col open"
  role="listbox"
  tabindex="0"
>
  <selected-item
    aria-selected="false"
    class="flex w-full items-center"
    role="option"
    tabindex="-1"
  >
    <!---->
    <items
      class="flex flex-wrap"
    >
      <!---->
      <!---->
      
      <!---->
       
      <p>
        One
      </p>
      <!---->
      <!---->
      <!---->
      
      <!---->
       
      <p>
        Two
      </p>
      <!---->
      
    </items>
    
    <!---->
     
    <rk-icon
      aria-label="opened"
      class=""
      data-tag="icon"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-opened"
      />
    </rk-icon>
    <!---->
  </selected-item>
   
  <scroll
    class="absolute z-10 flex w-full flex-col overflow-scroll menu"
    style="top: 0px;"
  >
    <rk-list
      aria-label="list"
      class=""
      role="listbox"
      tabindex="-1"
    >
      <!---->
       
      <rk-body>
        <!---->
        <rk-list-item
          aria-current="false"
          aria-selected="false"
          data-path="0"
          role="option"
        >
          <!---->
          <!---->
          
          <!---->
           
          <p>
            Three
          </p>
          <!---->
          <!---->
        </rk-list-item>
        
        <!---->
      </rk-body>
       
      <!---->
    </rk-list>
    <!---->
  </scroll>
  <!---->
</input-select>
`;
